core/state/state_object.go

```diff
@@ -341,6 +337,10 @@ func (s *stateObject) updateTrie() (Trie, error) {
 			return nil, err
 		}
 	}
+	// Short circuit if nothing changed, don't bother with hashing anything
+	if len(s.uncommittedStorage) == 0 {
+		return s.trie, nil
+	}
```

Comment:
Actually, I did this because we could have an account that is modified (e.g. balance change) and only reads storage.  And in that case, we would pick up the access list when we commit.

Reply:
> This will block until all prefetch tasks are done, which are needed for witnesses even for unmodified state objects.

My understanding of how the prefetcher works now is that:  as long as we have finished making `prefetch()` calls, subsequent calls to `trie()` will return with all pending requests completed (blocking if necessary).

There should be no other change we need to make for read-only accounts afaict.

Reply:
Ah right.  this will add overhead.  Here, we only need to resolve tries that will not be modified if we are in witness-building mode.  I will rewrite this part accordingly.